File: 0120-Triangle/lc0120.cc
// C++ solution.

class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        vector<int> dp(triangle.back());
        for (int nrow = triangle.size()-2; nrow >= 0; nrow--) {
            for (int i = 0; i <= nrow; i++) {
                dp[i] = min(dp[i], dp[i+1]) + triangle[nrow][i];
            }
        }
        return dp[0];
    }
};

File: 0120-Triangle/lc0120.go
// Go solution.

// DP method
func minimumTotal(triangle [][]int) int {
    tHeight := len(triangle)
    dp := make([]int, len(triangle[tHeight-1]))
    copy(dp, triangle[tHeight-1])

    // bottom-up iteration
    for i := tHeight - 2; i >= 0; i-- {
        for j := 0; j <= i; j++ {
            dp[j] = minInt(dp[j], dp[j+1]) + triangle[i][j]
        }
    }

    return dp[0]
}

func minInt(a, b int) int {
    if a > b {
        return b
    }
    return a
}